### PR TITLE
Major refactor to remove circular dependencies

### PR DESCRIPTION
**Pull request by corranwebster**
Key changes:

* break out BaseTraitHandler, TraitHandler, TraitType, CTrait, and TraitFactory
  into their own submodules.
* take the various trait converters/trait casters etc. and put them into their
  own submodule; plus simplify code via use of an informal `as_ctrait`
  interface so that we don't need as many imports.
* remove the dependency of List TraitType on items_event
* assorted other reshuffles and clean-up

As much as possible, if a name is available in a module, I've left it imported from its new location for backward compatibility.

### CODE REVIEW

**Comment by mdickinson** [comment]
@corranwebster Please can you fix the doc build? I'm seeing a couple of new warnings:
```
(traits-3.8) mirzakhani:docs mdickinson$ sphinx-build -b html source/ build/
Running Sphinx v2.3.1
making output directory... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 39 source files that are out of date
updating environment: [new config] 39 added, 0 changed, 0 removed
reading sources... [100%] traits_user_manual/testing                                                                                                                                                   
WARNING: autodoc: failed to import class 'HandleWeakRef' from module 'traits.trait_handlers'; the following exception was raised:
Traceback (most recent call last):
  File "/Users/mdickinson/.venvs/traits-3.8/lib/python3.8/site-packages/sphinx/util/inspect.py", line 230, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'traits.trait_handlers' has no attribute 'HandleWeakRef'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/mdickinson/.venvs/traits-3.8/lib/python3.8/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/Users/mdickinson/.venvs/traits-3.8/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/Users/mdickinson/.venvs/traits-3.8/lib/python3.8/site-packages/sphinx/ext/autodoc/__init__.py", line 1501, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/Users/mdickinson/.venvs/traits-3.8/lib/python3.8/site-packages/sphinx/util/inspect.py", line 246, in safe_getattr
    raise AttributeError(name)
AttributeError: HandleWeakRef

/Users/mdickinson/Enthought/ETS/traits/traits/trait_type.py:docstring of traits.trait_type.TraitType:20: WARNING: Field list ends without a blank line; unexpected unindent.
/Users/mdickinson/Enthought/ETS/traits/traits/trait_type.py:docstring of traits.trait_type.TraitType:37: WARNING: Field list ends without a blank line; unexpected unindent.
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] traits_user_manual/testing                                                                                                                                                    
generating indices...  py-modindexdone
highlighting module code... [100%] traits.util.trait_documenter                                                                                                                                        
writing additional pages...  searchdone
copying images... [100%] traits_user_manual/images/adaptation.png                                                                                                                                      
copying static files... ... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build succeeded, 3 warnings.

The HTML pages are in build.
```

**Comment by mdickinson** [comment]
I'm testing this with a couple of larger Traits-using projects.

**Comment by mdickinson** [comment]
> I'm testing this with a couple of larger Traits-using projects.

This got held up on finding that those larger projects currently fail against Traits master. I'm fixing those failures, and then will return to testing this PR branch.

**Comment by mdickinson** [comment]
No issues found running the test suites of those projects against this branch.

File: traits/trait_types.py [hunk 1]
```diff
@@ -3652,7 +3650,7 @@ def resolve_class(self, object, name, value):
 
 #: List of container type values; default value is ``[]``. This trait type is
 #: deprecated. Use ``List(This(allow_none=False))`` instead.
-ListThis = List(ThisClass)
+ListThis = List(This(allow_none=False))
```
Comment (mdickinson) [inline traits/trait_types.py]: Was this change intentional / necessary? Is this the result of `ThisClass` not being easily accessible?

Comment (corranwebster) [reply traits/trait_types.py]: Yes, it was because `ThisClass` was the only remaining direct import from `trait_handlers` and the comment suggested that the substitution would be equivalent.

**Comment by corranwebster** [comment]
I am fixing the docs in this PR; also would like to add some tests for the new `trait_converters` module.

**Comment by mdickinson** [comment]
@corranwebster Okay. Both those sound good to me.

**Comment by mdickinson** [comment]
Still LGTM. Thanks.
